### DISCUSSION TITLE
Help configuring postal to work over TLS

### DISCUSSION THREAD

**Discussion by davids355** [postalserver/postal#2689] [Help with using Postal] [answered]
Hi guys,
I have been trying all day to get postal working over TLS, to no avail.

In postal.yml I have added the following config - 
smtp_server:
  # Specify configuration to the Postal SMTP server
  port: 25
  tls_enabled: true
  tls_certificate_path: config/smtp.cert
  tls_private_key_path: config/smtp.key

Regarding the certificate files I have tried two approaches - 

linking the files from caddy - 
ln /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.domain/postal.domain.key smtp.key
ln /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.domain/postal.domain.crt smtp.cert

I also tried copying those files rather than linking them.

I also tried installing certbot and generating my own certificates and then copying those from the letsencrypt directory.

After each attempt I restarted postal, and I also tried restarting the entire server, but it did not work.

I also tried generating certificates using the key-type flag as I read that key for TLS has to be rsa - 
certbot certonly --standalone -d postal.domain `--key-type` `rsa`

But this did not work either.

Regardless of what I have tried above, when I test connection via gmass (https://www.gmass.co/smtp-test), I get the following output - 

Connected to smtp://postal.domain:25/?starttls=when-available
<< 220 postal.domain ESMTP Postal/IXSA9X
>> EHLO [ip]
<< 250-My capabilities are
<< 250-STARTTLS
<< 250 AUTH CRAM-MD5 PLAIN LOGIN
>> STARTTLS
<< 220 Ready to start TLS
ERROR: An error occurred while attempting to establish an SSL or TLS connection.
This usually means that the SSL certificate presented by the server is not trusted by the system for one or more of
the following reasons:
1. The server is using a self-signed certificate which cannot be verified.
2. The local system is missing a Root or Intermediate certificate needed to verify the server's certificate.
3. A Certificate Authority CRL server for one or more of the certificates in the chain is temporarily unavailable.
4. The certificate presented by the server is expired or invalid.
It is also possible that the set of SSL/TLS protocols supported by the client and server do not match.
See https://github.com/jstedfast/MailKit/blob/master/FAQ.md#SslHandshakeException for possible solutions.

Can someone help or point me in the right direction as to where I am going wrong.

**Comment by willpower232** [comment, marked as answer]
If you try `postal bash smtp` and `cat /config/smtp.cert` to confirm the contents then do you need to add a `/` at the start of both paths?

**Comment by davids355** [comment]
Ah fantastic such a small thing - I checked with cat and the contents were fine. Then I added that forward slash and hey-presto it works!
I did not realise/appreciate that the path was different in docker (Still learning about this stuff).

Now I am just wondering what is the correct way to have this working because I have tried so many things I am not sure what worked and what didnt. At the moment, I am now using a copy of certificates generated directly via letsencrypt (certbot) and also using the "key-type rsa" flag.

I am wondering if that was actually neccesary, or if the certificate generated by caddy should have worked?

And regardless, is it better that I create a symbolic link to one of those certificate files, so that when it renews everything will still work?

I did also see another thread where someone had written a script that copies the certificate files into the correct location every 3 months; is that neccesary? If so I will try that.

I will post links tomorrow if the above isn't clear, just rushing out of the office now.

I will also post an update with all the steps that I took to get this working, just in case anyone else comes across this thread (Because I have seen a lot of different threads about TLS and its took me all day to sift through and find a solution).

**Reply by airshell** [reply]
I am stuck with the same issue, it would be nice if you can mention the exact steps you followed to get postal working over SSL/TLS.  As the last step in the installation documentation mentions only about installing SSL via caddy.

Copying SSL files to config folder and configuring it correctly in postal.yml file, updating permissions etc. is missing from the available documentation.

I also tried adding "/" to file path in postal.yml file but it did not worked for me.

I previously had it configured with Nginx & certbot working well but then our hosting provider had changed IP's which resulted in certificate error. So I thought restart with fresh installation & SSL via caddy but now it does not connects over secure connection and giver error.

**Reply by davids355** [reply]
Don’t worry @airshell assuming you don’t get it fixed before tomorrow I’m definitely planning to post all the details here.

Im going to test tomorrow with both the caddy cert and the letsencrypt cert and try to establish what works and what doesn’t and then I’ll post everything I done here.

**Reply by willpower232** [reply]
As of https://github.com/postalserver/postal/pull/2126, you should be able to use ECDSA or RSA certificates with no issue. Others are apparently using the ones straight from caddy with no issue.

I think symbolic links would still work, might have to double check the permissions.

Regardless of how the files get into the docker containers, you will need to restart them when the certificates are replaced every 2/3 months.

I have experienced issues with long running processes (like docker) and symbolic links so it might just be easier to use copies of files.

**Comment by davids355** [comment]
@airshell I have just checked through my setup and I removed the certificate that I issued using letsencrypt and certbot and reverted back to the original certificates that were already generated by caddy and they worked as well. So my only issue was not adding the forward slash at the beginning of the certificate paths in the postal config.

This is a summary of what I done to get it working - 

Remove any keys from config folder if they exist - 
cd /opt/postal/config/
rm smtp.cert smtp.key

Copy on the certificates from caddy - 
cp /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.com/postal.mydomain.com.key smtp.key
cp /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.com/postal.mydomain.com.crt smtp.cert

Obviously whatever name you give the certs has to match the postal config (in my case I call them smtp.key and smtp.cert).

edit postal config - 
nano /opt/postal/config/postal.yml

Add this content - 
smtp_server:
  port: 25
  tls_enabled: true
  tls_certificate_path: /config/smtp.cert
  tls_private_key_path: /config/smtp.key

Note - I found two things here were important for me - 
1. someone correct me if I am wrong but I think that having the correct indentation in this config file is important?
2. I had to add a slash at the beginning of the cert paths as I understand that in the docker container the config folder is mounted at root (/config).

Restart postal - 
postal stop
postal start

For me, it worked at this point and I don't beleive that I have done anything else to get it working - it was very simple, just important that cert names are correct and that the path is correct in postal config.

I then tested with gmass smtp tool and it shows full connection with tls.

If its not working for you, some troubleshooting - 

As @willpower232 mentioned above, ssh into the docker container and make sure you can see and read the certs - 

postal bash smtp
cat /config/smtp.cert
cat /config/smtp.key

You should see their contents, then you know they are present, docker container can see them and they hold the correct info.

I also followed this guide for creating a script that will check for the certificates changing (being renewed) and then re-copy them in and restart postal - 
https://github.com/postalserver/postal/discussions/2673

In here, the person mentions about updating permissions for the cert files to make sure postal can read them. You could try that as well.

Try the above and let me know how you get on. If its not working then perhaps between us we can work out some more troubleshooting steps or think about what might be wrong.

**Reply by airshell** [reply]
Thanks @davids355 @willpower232 

Here are the steps I followed.

I removed existing SSL related files from config folder -
cd /opt/postal/config/
rm smtp.cert smtp.key

And then tried to copy them again but this time the files are not getting copied. I used below command that you shared.

Copy on the certificates from caddy -
cp /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.com/postal.mydomain.com.key smtp.key
cp /opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.com/postal.mydomain.com.crt smtp.cert

I tried to `ls /config/` to view files in config folder but it does not show the files, image below for reference.

![ssl-files-not-copied](https://github.com/postalserver/postal/assets/149798684/3005747f-27ab-41df-ab92-b8e58ff705fc)

I guess this might be related to config directory permission and symbolic links as mentioned by @willpower232 , since I remember during installation I used below command to create symlink.
`sudo ln -s /usr/local/bin/docker-compose /usr/bin/docker-compose`

@willpower232 can you plz let me know what should be the `/config/` directory permission so I can update it & copy the files. 
As you have previous experience with symbolic links, after adjusting permission I will try again to copy SSL files from caddy folder as I see SSL files present in `/opt/postal/caddy-data/caddy/certificates/acme-v02.api.letsencrypt.org-directory/postal.mydomain.com/`.

Once the files get copied I plan to adjust permissions of SSL files to ensure Postal can read the certificates.
`chmod o+r /opt/postal/config/smtp.*`

And create a task for Monitoring and Updating Postal Certificates as mentioned in #2673

Thanks again I really appreciate your help.

**Reply by willpower232** [reply]
The Dockerfile creates a user called postal https://github.com/postalserver/postal/blob/2f62baa238fc1102706ee4acf079b7a876b05283/Dockerfile#L29

Not sure how that evaluates in numeric linux user IDs

**Comment by davids355** [comment]
@airshell to troubleshoot I would check locally first - after running those copy commands, navigate to your local postal directory which should be /opt/postal/config/
then either dir or ls to show files in there, and see if the cert files are present. If they are not, then just make sure you are in the /opt/postal/config directory before running the copy commands. thats the first step.

Once thats done, I imagine you might have to restart postal (or possibly the entire server) so that the files become available in the docker.
So once thats restarted, then ssh into the docker container and check if the cert files are present in /config folder (Which is actually reading the contents from /opt/postal/config on the local system).

Excuse if some of my terminology is incorrect because I am still learning about many of this technologies and principals :-) 

@willpower232 may correct some of the above but its a starting point in the meantime.
